File: linked_list/java/LinkedList.java
public class LinkedList<T> {
    public static void main(String[] args) {
        LinkedList<String> list = new LinkedList<String>("head");
        list.add("second");
        list.add("third");
        list.add("last");
        System.out.println(list.length()); // should be 4
        System.out.println(list.remove("third")); // should print "third"
        System.out.println(list.length()); // should be 3
    }

    private Node<T> head;

    public LinkedList(T head) {
        this.head = new Node<T>(head);
    }

    public LinkedList() {
    }

    /**
     * Adds a value to the end of the list
     *
     * @param val value to be added
     * @return true if add succeeds; false otherwise
     */
    public boolean add(T val) {
        Node<T> tmp = this.head;
        if (tmp == null) {
            this.head = new Node<T>(val);
            return true;
        }

        while (tmp.next() != null) {
            tmp = tmp.next();
        }

        tmp.setNext(new Node<T>(val));
        return true;
    }

    /**
     * Returns the value that was removed if the requested value
     * was found. Otherwise returns null if the value doesn't exist
     * in the list
     * @param val
     * @return T
     */
    public T remove(T val) {
        if (this.head != null) {
            if (this.head.getValue() == val) {
                Node<T> trash = this.head;
                this.head = this.head.next();
                trash.setNext(null);
                return trash.getValue();
            }
        } else {
            return null;
        }

        Node<T> prev = this.head;
        Node<T> tmp = prev.next();
        while (tmp != null) {
            if (tmp.getValue() == val) {
                prev.setNext(tmp.next());
                tmp.setNext(null);
                return tmp.getValue();
            }
            tmp = tmp.next();
            prev = prev.next();
        }

        return null;
    }

    public int length() {
        int len = 0;
        Node<T> tmp = this.head;
        while (tmp != null) {
            len++;
            tmp = tmp.next();
        }

        return len;
    }

    private class Node<T> {
        private Node<T> nextNode;
        private T value;

        public Node(T val) {
            this.value = val;
        }

        public Node<T> next() {
            return this.nextNode;
        }

        public void setNext(Node<T> next) {
            this.nextNode = next;
        }

        public T getValue() {
            return this.value;
        }
    }
}

File: linked_list/Python/linked_list.py
#!/usr/bin/env python3


class NoSuchNodeException(Exception):
    """Exception class"""
    pass


class Node:
    """
    Class for the structure of node in Linked List

    :param data:    Data value for node
    :param next_node:   Next node reference
    """

    def __init__(self, data=None, next_node=None):
        self.data = data
        self.next_node = next_node

    def compare_to(self, node):
        """
        Method to compare node data

        :return:    Returns True if node data matches
        """
        return node.data == self.data


class LinkedList:
    """
    Class for the Linked List
    :param head:    Reference of the head.
    :param list_size:   List size.
    """

    def __init__(self, head=None, list_size=0):
        self.head = head
        self.list_size = list_size

    def size(self):
        """
        This methods returns the size of the linked list.

        :return:    The size of the linked list.
        """
        return self.list_size

    def is_empty(self):
        """
        This methods determines whether the list is empty.
        :return:    True if the list is empty, otherwise False.
        """
        return self.list_size == 0

    def get_first(self):
        """
        This method returns the first element of the list.

        :return:    The first element of the list.
        """
        if self.is_empty():
            raise NoSuchNodeException()

        return self.head.data

    def getNode(self, i):
        """
        This method takes an index and finds the node residing.

        :raises:    NoSuchNodeException if no such element.
        :param i:   The index of the node.
        :return:    The ith node of the list.
        """
        if i < 0 or i > self.list_size - 1:
            raise NoSuchNodeException()

        current = 0
        p = self.head

        while current < i:
            p = p.next_node
            current += 1

        return p

    def add_first(self, data):
        """
        This method adds a node to the front of the list.

        :param data:    The data of the node to add.
        """
        node = Node(data, self.head)
        self.head = node

        self.list_size += 1

    def add_last(self, data):
        """
        This method adds a node to the back of the list.

        :param data:    The data of the node to add.
        """
        node = Node(data, None)

        if self.is_empty():
            self.head = node
        else:
            tail = self.getNode(self.list_size - 1)
            tail.next_node = node

        self.list_size += 1

    def remove_first(self):
        """
        This method removes the first element of the list.

        :return:    The value of the first element.
        """
        if self.is_empty():
            raise NoSuchNodeException()

        tmp_val = self.head.data
        self.head = self.head.next_node
        self.list_size -= 1
        return tmp_val

    def remove_last(self):
        """
        This method removes the last element of the list.

        :return:    The value of the removed element.
        """
        if self.is_empty():
            raise NoSuchNodeException()

        tail = self.getNode(self.list_size - 1)
        tail_data = tail.data

        if self.list_size == 1:
            self.head = None
        else:
            before_tail = self.getNode(self.list_size - 2)
            before_tail.next_node = None

        self.list_size -= 1
        return tail_data

    def print_list(self):
        """
        Method to print the list in a newline after each element.
        """
        p = self.head
        i = 0

        while i < self.size():
            print(p.data)
            i += 1
            p = p.next_node


def main():
    myList = LinkedList()
    myList.add_first(10)
    myList.add_first(20)
    myList.add_first(30)
    print("Current List after adding 3 elements: ")
    myList.print_list()

    myList.remove_last()
    print("Current List after removing last element: ")
    myList.print_list()


if __name__ == "__main__":
    main()
